src/app/icd/ICDClientInfoPersistentStorage.h

```diff
@@ -0,0 +1,165 @@
+ * for storing and retrieving ICD client information in persistent storage.
+ */
+class ICDClientInfoPersistentStorage
+{
+public:
+    /**
+     * Struct to store information about ICD Client Info.
+     */
+    struct ICDClientInfo
+    {
+        NodeId mPeerNodeId                 = kUndefinedNodeId;
+        uint32_t mStartICDCounter          = 0;
+        uint32_t mOffset                   = 0;
+        uint64_t mMonitoredSubject         = static_cast<uint64_t>(0);
+        Crypto::Aes128KeyHandle mSharedKey = Crypto::Aes128KeyHandle();
+
+        bool IsValid() { return mPeerNodeId != kUndefinedNodeId; }
+    };
+
+    /**
+     * Struct per identity representing persistent storage for ICD client information and key store
+     */
+    struct Storage
+    {
+        FabricIndex mFabricIndex                      = kUndefinedFabricIndex;
+        PersistentStorageDelegate * mpClientInfoStore = nullptr;
+        Crypto::SessionKeystore * mpKeyStore          = nullptr;
+        void Init(FabricIndex aFabricIndex, PersistentStorageDelegate * apClientInfoStore, Crypto::SessionKeystore * apKeyStore)
+        {
+            mFabricIndex      = aFabricIndex;
+            mpClientInfoStore = apClientInfoStore;
+            mpKeyStore        = apKeyStore;
+        }
+        Storage() {}
+        ~Storage() {}
+        Storage(Storage && aOther)
+        {
+            mFabricIndex             = aOther.mFabricIndex;
+            mpClientInfoStore        = aOther.mpClientInfoStore;
+            mpKeyStore               = aOther.mpKeyStore;
+            aOther.mpClientInfoStore = nullptr;
+            aOther.mpKeyStore        = nullptr;
+            aOther.mFabricIndex      = kUndefinedFabricIndex;
+        }
+        Storage & operator=(const Storage & aOther) = default;
+        bool IsValid() { return mFabricIndex != kUndefinedFabricIndex && mpClientInfoStore != nullptr && mpKeyStore != nullptr; }
+    };
+
+    using ICDClientInfoIterator = CommonIterator<ICDClientInfo>;
+
+    virtual ~ICDClientInfoPersistentStorage(){};
+
+    /**
+     * Iterate through persisted ICD Client Info
+     *
+     * @return A valid iterator on success. Use CommonIterator accessor to retrieve ICDClientInfo
+     */
+    virtual ICDClientInfoIterator * IterateICDClientInfo() = 0;
+
+    /**
+     * Generate ICD shared key handle, update clientInfo with key handle, and store it in storage.
+     *
+     * @param aICDClientInfo the ICD Client information to be updated with keyData and be saved
+     * @param aFabricIndex the fabric indeX
+     * @param aKeyData raw key data
+     *
+     */
+    virtual CHIP_ERROR CreateICDInfoEntry(ICDClientInfo & aICDClientInfo, FabricIndex aFabricIndex, const ByteSpan & aKeyData) = 0;
+
+    /**
+     * Delete ICD Client persistent information by node ID, fabric index.
+     */
+    virtual CHIP_ERROR DeleteEntry(NodeId aNodeId, FabricIndex aFabricIndex) = 0;
```

Comment:
Will the SDK be calling this?  Under what conditions?

Reply:
updated, thanks